### PR TITLE
[Foundation] Added NSTask.h

### PR DESCRIPTION
**Pull request by PoomSmart**

### CODE REVIEW

**Comment by leptos-null** [comment]
I'm seeing these are marked `Copyright (c) 1996-2019, Apple Inc. All rights reserved.`; are these available under a sufficiently permissive license for this repo?

File: Foundation/NSTask.h [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+//   { Name = NSTaskDidTerminateNotification; object = task; }
+//
+
+- (instancetype)init NS_DESIGNATED_INITIALIZER;
+
+// these methods can only be set before a launch
+@property (nullable, copy) NSURL *executableURL API_AVAILABLE(macos(10.13)) API_UNAVAILABLE(ios, watchos, tvos);
+@property (nullable, copy) NSArray<NSString *> *arguments;
+@property (nullable, copy) NSDictionary<NSString *, NSString *> *environment; // if not set, use current
+@property (nullable, copy) NSURL *currentDirectoryURL API_AVAILABLE(macos(10.13)) API_UNAVAILABLE(ios, watchos, tvos);
```
Comment (leptos-null) [inline Foundation/NSTask.h:32]: should these availability attributes be updated?
```suggestion
@property (nullable, copy) NSURL *currentDirectoryURL API_AVAILABLE(macos(10.13));
```

Comment (leptos-null) [reply Foundation/NSTask.h]: the "became" available attributes look valid to me. should they be kept in?

Comment (PoomSmart) [reply Foundation/NSTask.h]: Do we want the headers to be compatible with macOS too?

**Comment by uroboro** [comment]
> I'm seeing these are marked `Copyright (c) 1996-2019, Apple Inc. All rights reserved.`; are these available under a sufficiently permissive license for this repo?

Is there any SDK component that is?

File: Foundation/NSTask.h (hunk 1 above)
Comment (uroboro) [reply Foundation/NSTask.h]: It'd be a nice to have but not a requirement, I think

**Review by uroboro — APPROVED** [review]

**Comment by leptos-null** [comment]
@uroboro some headers are shipped in the SDK that are part of an open source project, e.g. xnu, IOKit

**Review by kabiroberai — CHANGES_REQUESTED** [review]
Can we be sure that this header won't conflict with the actual NSTask.h header in SDKs that have that? We shouldn't be breaking any macOS project (of which I know a few) with this one change.

**Comment by kabiroberai** [comment]
@leptos-null to my knowledge header files can't be copyrighted; see https://softwareengineering.stackexchange.com/a/216480. To elaborate on Oracle v Google, the case went all the way up to the US Supreme Court which ruled in April 2021 that API interfaces (like headers) are covered under fair use.

**Comment by kirb** [comment]
I really feel this is best to go in the patched SDKs, since it toes the line of what is and isn’t copyrightable too much for my liking, even if Oracle v. Google seems to indicate it should be fine. Would rather keep that sort of thing contained to the sdks repo so in the worst case scenario, only that repo is lost.

If we do still choose to merge it here, we should wrap the header in

```objc
#if __has_include(<Foundation/NSTask.h>)
#include_next <Foundation/NSTask.h>
#else

…

#endif
```

so we defer to the SDK header when it exists. Also, please annotate it in LICENSE.md.

**Comment by kabiroberai** [comment]
Good point @kirb, I'd rather it be in theos/sdks too. For that matter, why do we have any headers from system frameworks (class-dump'd or not) here?

**Comment by kirb** [comment]
It’s really mostly just because the headers repo predates the sdks repo. It’s also far easier to maintain this repo since it applies regardless of SDK version, or if you even have a patched SDK set up at all. However, where we can keep patched-up official headers current with SDK changes in an automated fashion, i.e. by grabbing headers from MacOSX SDKs or patching availability checks, we absolutely should do it in the sdks repo.

**Comment by kabiroberai** [comment]
Makes sense. The reason I ask is that moving patches to public frameworks into the SDKs repo would automatically make them compatible with Swift as well. As it stands, we'd need to modify the system framework's modulemap to add new headers – unless it's possible to extend a modulemap outside its initial declaration.

**Comment by PoomSmart** [comment]
Be it in sdks it is.

**Pull request closed without merging**